linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/ClassEmitter.scala

```diff
@@ -805,32 +795,22 @@ private[emitter] final class ClassEmitter(sjsGen: SJSGen) {
             genIsInstanceOfHijackedClass(obj, className)
 
           case _ =>
-            var test = if (tree.kind.isClass) {
-              genIsInstanceOfClass(obj, className)
-            } else {
-              !(!(
-                  genIsScalaJSObject(obj) &&
-                  genIsClassNameInAncestors(className,
-                      obj DOT "$classData" DOT "ancestors")
-              ))
+            val baseTest = {
+              if (tree.kind.isClass) {
+                genIsInstanceOfClass(obj, className)
+              } else {
+                !(!(
+                    genIsScalaJSObject(obj) &&
+                    genIsClassNameInAncestors(className,
+                        obj DOT "$classData" DOT "ancestors")
+                ))
+              }
             }
 
-            def typeOfTest(typeString: String): js.Tree =
-              js.UnaryOp(JSUnaryOp.typeof, obj) === js.StringLiteral(typeString)
-
-            if (isAncestorOfString)
-              test = test || typeOfTest("string")
-            if (isAncestorOfHijackedNumberClass) {
-              test = test || typeOfTest("number")
-              if (useBigIntForLongs)
-                test = test || genCallHelper("isLong", obj)
+            globalKnowledge.hijackedDescendants(className).foldLeft(baseTest) {
+              case (test, hijackedClass) =>
+                test || genIsInstanceOfHijackedClass(obj, hijackedClass)
             }
```

Comment:
This is the only place where we use the actual *list* of `hijackedDescendants`, and it is the one causing the size/performance regression. I'm not too worried about `j.l.Number` itself, but more about interfaces like `Comparable`. I think we could easily reclaim the performance by eliminating from the list the type test that are subsumed by another one in the list. Or, more directly, if the list contains `j.l.Double`, we can get rid of all the other boxed number classes (except `j.l.Long`).

We could also ignore `j.l.Long` here unless `useBigIntForLongs` is true.

None of those changes require anything but local knowledge, so this should be feasible here. We can protect this slightly more computation by a `isAncestorOfHijackedClass(className)`.